src/main/flight/autopilot.c

```diff
@@ -22,30 +22,97 @@
 
 #include "platform.h"
 #include "build/debug.h"
+#include "common/filter.h"
 #include "common/maths.h"
+#include "fc/rc.h"
+#include "fc/runtime_config.h"
 
 #include "flight/imu.h"
-#include "flight/pid.h"
 #include "flight/position.h"
 #include "rx/rx.h"
+#include "sensors/gyro.h"
+#include "sensors/compass.h"
 
 #include "autopilot.h"
 
 #define ALTITUDE_P_SCALE  0.01f
 #define ALTITUDE_I_SCALE  0.003f
 #define ALTITUDE_D_SCALE  0.01f
 #define ALTITUDE_F_SCALE  0.01f
+#define POSITION_P_SCALE  0.001f
+#define POSITION_I_SCALE  0.0005f
+#define POSITION_D_SCALE  0.002f
+#define POSITION_A_SCALE  0.002f
 
 static pidCoefficient_t altitudePidCoeffs;
+static pidCoefficient_t positionPidCoeffs;
+
 static float altitudeI = 0.0f;
 static float throttleOut = 0.0f;
 
+typedef struct {
+    float gpsDataIntervalS;
+    float gpsDataFreqHz;
+    float sanityCheckDistance;
+    bool isStartingNS;
+    bool isStartingEW;
+    float peakInitialGroundspeed;
+    float lpfCutoff;
+    float pt1Gain;
+    bool sticksActive;
+    float previousDistanceNS;
+    float previousDistanceEW;
+    float previousVelocityNS;
+    float previousVelocityEW;
+    float previousPidSumNS;
+    float previousPidSumEW;
+    float integralNS;
+    float integralEW;
```

Comment:
This is just crazy. Please  use vectors + vector math. Separating coordinates would only lead to problem and would be hard to remove in future.

Reply:
Unfortunately for me vector math is too complex.  

However, already I found a significant improvement in precision by avoiding the bearing vector method and going to direct coordinate distances.  The GPS gives us XY position.  A vector approach requires calculation of distances and angles between points, converting them into vectors, then doing PIDs on vectors, and I don't know any easy way to do PID controllers on angles.  Each conversion from position values to vectors is slightly imprecise and requires a lot of maths.  

The XY approach directly uses the numbers we receive from the GPS with no conversions or loss of precision.  The PIDs are easy to calculate, and the process is easy on the CPU, with only one sin and cos transformation required. Having just one PID controller for latitude, and another for longitude is conceptually simple.  And if we have a glitch on only one axis, that changes the vector length; if we rotate the vector, then we can transform a glitch on one axis to another.  Since we actually get inputs in latitude and longitude position, this separate PID approach for each axis isolates glitches on one axis, completely, from influencing the other, in terms of their PIDs.  This is important for the filtered D where a spike will have significant duration and is best isolated to the axis with the spike.

If someone else wants to do this in vectors and can show either that it performs better, or is easier on the CPU, then I'll welcome that PR.  At this point I want to keep this PR as it is.